File: eBPF_Marine/pps_vide.py
from bcc import BPF
import time
import signal
import sys
import argparse
import csv
import os

device = "veth-main"
b = None

# Création du dossier pour les logs CSV
os.makedirs("./pps_vide_data", exist_ok=True)

csv_file = open("./pps_vide_data/pps_vide_log.csv", mode='w', newline='')
csv_writer = csv.writer(csv_file)
csv_writer.writerow(["timestamp", "var"])

def cleanup(signum=None, frame=None):
    print("\n[!] Nettoyage...")
    try:
        if b:
            b.remove_xdp(device)
            print(f"[✓] XDP détaché de {device}")
    except Exception as e:
        print(f"[!] Erreur nettoyage: {e}")
    csv_file.close()
    print("Données enregistrées dans './pps_vide_data/pps_vide_log.csv'")
    sys.exit(0)

def clear_maps():
    b["ip_count_map"].clear()
    b["ip_flood_count_map"].clear()
    b["ip_timestamp_map"].clear()
    b["ip_blacklist_map"].clear()
    b["port_blacklist_map"].clear()
    b["pps"].clear()
    print("[INFO] Maps eBPF nettoyées")

signal.signal(signal.SIGINT, cleanup)
signal.signal(signal.SIGTERM, cleanup)

def main():
    global b

    parser = argparse.ArgumentParser(description="Mesure PPS XDP")
    parser.add_argument("--avg", type=int, default=5, help="")
    args = parser.parse_args()

    b = BPF(src_file="pps.c")
    fn = b.load_func("xdp_pps", BPF.XDP)
    b.attach_xdp(device, fn, 0)
    clear_maps()
    print(f"Programme XDP attaché à {device}")

    counter = b.get_table("pps")
    key = 0

    print("Mesure du nombre de paquets traités par seconde :\n")

    while True:
        time.sleep(1)
        total = 0
        for cpu_val in counter[key]:
            total += cpu_val
        #print(f"{total} paquets/s")
        del counter[key]

        ts = int(time.time())
        csv_writer.writerow([ts, total])

if __name__ == "__main__":
    main()

File: eBPF_Marine/pps_vide.c
#include <uapi/linux/bpf.h>
#include <linux/in.h>
#include <linux/if_ether.h>
#include <linux/ip.h>
#include <linux/tcp.h>
#include <linux/udp.h>
#include <asm/byteorder.h> // Pour ntohl()

// Maps pour traquer les IPs, timestamps, blacklists et ports bloqués
BPF_HASH(ip_flood_count_map, u32, u64); // valeur = nombre de paquets reçus en l'espace d'une seconde
BPF_HASH(ip_count_map, u32, u64); // valeur = nombre total de paquets envoyés par cette ip
BPF_HASH(ip_timestamp_map, u32, u64);
BPF_HASH(ip_blacklist_map, u32, u8);
BPF_HASH(port_blacklist_map, u16, u8);
BPF_PERCPU_ARRAY(pps, u64, 1);


int xdp_pps_vide(struct xdp_md *ctx) {
    // Mesure pps (=paquets traités par seconde)
    u32 key = 0;
    u64 *value = pps.lookup(&key);
    if (value) {
        __sync_fetch_and_add(value, 1);
    }

    return XDP_PASS;
}
